modusocket settimeout on unbound sockets.

iabdalkader opened the pull request:
* For extended state socket, if `settimeout()` is called before a NIC is bound, save the timeout until the NIC is bound.
* Allow setting timeout on unbound sockets.
* Initialise accepted socket state.
* Use socket timeout preset in modusocket.

iabdalkader committed:
extmod/modusocket.c: Allow setting timeout on unbound sockets.

iabdalkader committed:
extmod/network_ninaw10.c: Use socket timeout preset in modusocket.

dpgeorge commented:
For the `accept` call, should the setting of timeout to 0 be moved to `modusocket.c` as well?

dpgeorge commented:
Regarding `accept` timeout settings, see related #2948.

iabdalkader commented:
> For the accept call, should the setting of timeout to 0 be moved to modusocket.c as well?

Yes, the extended state, including the timeout, should be initialized in `accept()` for `socket2`, and you can override it in the driver if needed. I also noticed that proto, type, domain etc.. are not set for socket2, should we copy those from the server socket ?

>Regarding accept timeout settings, see related

I'm not sure if it should be 0 or inherited, I typically set the timeout on sockets returned from accept because they are usually different. For example I want the server socket to be blocking, and client/accept sockets to timeout:
```Python
    server.setblocking(True)
    ....
    client, addr = s.accept()
    client.settimeout(5.0)
```

Let me know if timeout should be copied or set to 0.

dpgeorge commented:
> I also noticed that proto, type, domain etc.. are not set for socket2, should we copy those from the server socket ?

Yes I think they should be copied so at least they have valid values.

> Let me know if timeout should be copied or set to 0.

I think set it to 0?  The proper way to use accepted sockets would be to initialise their timeout to what you need (ie not rely on the default).

iabdalkader committed:
extmod/modusocket.c: Initialise accepted socket state.

iabdalkader commented:
> Yes I think they should be copied so at least they have valid values.

Rebased and fixed. I split the accepted socket init into two parts, before the NIC's accept call (where timeout is set to 0) and after. This way a driver can override the timeout for accepted sockets, if needed for some reason, but let me know if you want to move it after NIC's accept call. 

> The proper way to use accepted sockets would be to initialise their timeout to what you need (ie not rely on the default).

Probably should document this somewhere.

dpgeorge commented on extmod/modusocket.c:
```diff
@@ -155,6 +168,10 @@ STATIC mp_obj_t socket_accept(mp_obj_t self_in) {
+    // set the same address family, socket type and protocol as parent.
+    socket2->domain = self->domain;
+    socket2->type = self->type;
+    socket2->proto = self->proto;
```
I think this should be moved up, to before the `->accept()` call.  This is because `wiznet5k_socket_accept()` sets these values and does something tricky with `self->domain`, and I think this change will break that.

iabdalkader replied:
Fixed.

dpgeorge commented:
Thanks for fixing.  Merged in 67420de4f4faa214de853dc70ef307d3571bfc28 through f2ccf87e0b139d7e0402e9f4a615cedfeac7a11e

iabdalkader commented:
@robert-hh settimeout issue should be fixed now.

iabdalkader commented:
> I was wondering whether select() on sockets work

Not without `ioctl` which may be possible to implement.
